src/vos/vos_ts.h

```diff
@@ -419,82 +425,91 @@ vos_ts_evict(uint32_t *idx, uint32_t type)
 
 /** Allocate thread local timestamp cache.   Set the initial global times
  *
- * \param	ts_table[in,out]	Thread local table pointer
+ * \param[in,out]	ts_table	Thread local table pointer
  *
- * \return	-DER_NOMEM	Not enough memory available
- *		0		Success
+ * \return		-DER_NOMEM	Not enough memory available
+ *			0		Success
  */
 int
 vos_ts_table_alloc(struct vos_ts_table **ts_table);
 
 
 /** Free the thread local timestamp cache and reset pointer to NULL
  *
- * \param	ts_table[in,out]	Thread local table pointer
+ * \param[in,out]	ts_table	Thread local table pointer
  */
 void
 vos_ts_table_free(struct vos_ts_table **ts_table);
 
-/** Update the low read timestamp, if applicable
- *
- *  \param	entry[in,out]	Entry to update
- *  \param	epoch[in]	Update epoch
- */
-static inline void
-vos_ts_update_read_low(struct vos_ts_entry *entry, daos_epoch_t epoch)
-{
-	if (entry == NULL)
-		return;
-	entry->te_ts_rl = MAX(entry->te_ts_rl, epoch);
-}
-
-/** Update the high read timestamp, if applicable
- *
- *  \param	entry[in,out]	Entry to update
- *  \param	epoch[in]	Update epoch
- */
-static inline void
-vos_ts_update_read_high(struct vos_ts_entry *entry, daos_epoch_t epoch)
-{
-	if (entry == NULL)
-		return;
-	entry->te_ts_rh = MAX(entry->te_ts_rh, epoch);
-}
-
 /** Allocate a timestamp set
  *
- * \param	ts_set[in,out]	Pointer to set
- * \param	flags[in]	Operations flags
- * \param	akey_nr[in]	Number of akeys in operation
+ * \param[in,out]	ts_set	Pointer to set
+ * \param[in]		flags	Operations flags
+ * \param[in]		akey_nr	Number of akeys in operation
+ * \param[in]		tx_id	Optional transaction id
  *
  * \return	0 on success, error otherwise.
  */
 int
 vos_ts_set_allocate(struct vos_ts_set **ts_set, uint64_t flags,
-		    uint32_t akey_nr);
+		    uint32_t akey_nr, uuid_t *tx_id);
 
 /** Upgrade any negative entries in the set now that the associated
  *  update/punch has committed
  *
- *  \param	ts_set[in]	Pointer to set
+ *  \param[in]	ts_set	Pointer to set
  */
 void
 vos_ts_set_upgrade(struct vos_ts_set *ts_set);
 
 /** Free an allocated timestamp set
  *
- * \param	ts_set[in]	Set to free
+ * \param[in]	ts_set	Set to free
  */
 static inline void
 vos_ts_set_free(struct vos_ts_set *ts_set)
 {
 	D_FREE(ts_set);
 }
 
+/** Update the low timestamp if the new read is newer
+ *
+ * \param[in]	entry		The timestamp entry
+ * \param[in]	read_time	The new read timestamp
+ * \param[in]	tx_id		The uuid of the new read
+ */
+static inline void
+vos_ts_rl_update(struct vos_ts_entry *entry, daos_epoch_t read_time,
+		 const uuid_t tx_id)
+{
+	if (entry == NULL || read_time <= entry->te_ts_rl)
```

Comment:
If read_time = te_ts_rl but tx_id != te_tx_rl, then we should assign tx_id to te_tx_rl, no?

Reply:
yes, perhaps we should.   Though it's a little unclear to me.   If two transactions have same timestamp and both try to modify the same key, one will be asked to restart.

Reply:
Yeah, this requires a little deduction from the rules in src/vos/README.md. Consider:

    TX1 reads  value x at 10: x.low = {10, TX1}
    TX2 reads  value x at 10: x.low = {10, TX2} <-- required to reject next line
    TX1 writes value x at 10: reject due to TX ID differences